File: typescript/src/model/types/wdl-array-type.ts
/** Array type nodes. */
import { WdlType, WdlTypeComponentType } from './wdl-type.js';

/** Models an array type such as `Array[Int]` or `Array[String]+`. */
export class WdlArrayType extends WdlType {
  /** Creates an array type from its member type, non-empty flag, and optional flag. */
  public constructor(
    private arrayMemberType?: WdlType,
    private nonEmpty = false,
    optional = false,
  ) {
    super(optional);
  }

  /** Returns the broad type family for this node. */
  public componentType(): WdlTypeComponentType {
    return WdlTypeComponentType.ARRAY;
  }

  /** Returns the member type stored in the array. */
  public memberType(): WdlType | undefined {
    return this.arrayMemberType;
  }

  /** Sets the member type stored in the array. */
  public setMemberType(arrayMemberType: WdlType | undefined): void {
    this.arrayMemberType = arrayMemberType;
  }

  /** Returns whether the array has the WDL non-empty (`+`) marker. */
  public isNonEmpty(): boolean {
    return this.nonEmpty;
  }

  /** Sets whether the array has the WDL non-empty (`+`) marker. */
  public setNonEmpty(nonEmpty: boolean): void {
    this.nonEmpty = nonEmpty;
  }
}

File: java/src/main/java/com/myriad/wdl/model/types/WdlArrayType.java
package com.myriad.wdl.model.types;

/**
 * Array type node.
 *
 * <p>This models WDL array types such as {@code Array[Int]} and {@code Array[String]+}. The
 * {@code nonEmpty} flag represents the `+` suffix used by the specification for non-empty arrays.
 */
public final class WdlArrayType extends WdlType {
  private WdlType arrayMemberType;
  private boolean nonEmpty;

  public WdlArrayType() {
    super();
  }

  public WdlArrayType(WdlType arrayMemberType, boolean nonEmpty, boolean optional) {
    super(optional);
    this.arrayMemberType = arrayMemberType;
    this.nonEmpty = nonEmpty;
  }

  @Override
  public ComponentType componentType() {
    return ComponentType.ARRAY;
  }

  /** Returns the array member type. */
  public WdlType memberType() {
    return arrayMemberType;
  }

  /** Sets the array member type. */
  public void setMemberType(WdlType arrayMemberType) {
    this.arrayMemberType = arrayMemberType;
  }

  /** Returns whether the type carries the WDL non-empty array marker (`+`). */
  public boolean isNonEmpty() {
    return nonEmpty;
  }

  /** Sets whether the type carries the WDL non-empty array marker (`+`). */
  public void setNonEmpty(boolean nonEmpty) {
    this.nonEmpty = nonEmpty;
  }
}
